src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -287,35 +286,18 @@ private Solution GetLSPSolution(Solution workspaceSolution)
         {
             var workspace = workspaceSolution.Workspace;
 
-            var recalculateSolution = false;
-            // if we don't have a cached LSP view of the world, we need to create a new one
-            if (!_lastLSPSolution.ContainsKey(workspace))
+            // If we have a cached solution we can use it, unless the workspace solution it was based on
+            // is not the current one. 
+            if (!_lspSolutionCache.TryGetValue(workspace, out var cacheInfo) ||
+                workspaceSolution != cacheInfo.workspaceSolution)
             {
-                recalculateSolution = true;
-            }
-
-            // if the workspace has changed we need to recalculate, and save the current state as our starting point
-            if (!_lastWorkspaceSolution.TryGetValue(workspace, out var lastSolution) || lastSolution != workspaceSolution)
-            {
-                recalculateSolution = true;
-                _lastWorkspaceSolution[workspace] = workspaceSolution;
-            }
+                var lspSolution = GetSolutionWithReplacedDocuments(workspaceSolution);
+                                _lspSolutionCache[workspace] = (workspaceSolution, lspSolution);
```

Comment:
not a fan of this indentation.

Reply:
Rest assured this was not intentional. Must be ghosts